File: src/main/java/me/syari/ss/discord/internal/entities/Role.java
package me.syari.ss.discord.internal.entities;

import me.syari.ss.discord.api.ISnowflake;
import me.syari.ss.discord.internal.JDAImpl;
import me.syari.ss.discord.internal.utils.cache.SnowflakeReference;
import org.jetbrains.annotations.NotNull;

import java.time.OffsetDateTime;

public class Role implements ISnowflake {
    private final long id;
    private final SnowflakeReference<Guild> guild;

    private String name;

    public Role(long id, @NotNull Guild guild) {
        this.id = id;
        JDAImpl api = guild.getJDA();
        this.guild = new SnowflakeReference<>(guild, api::getGuildById);
    }

    @NotNull
    public String getName() {
        return name;
    }

    @NotNull
    public String getAsMention() {
        return "<@&" + getId() + '>';
    }

    @Override
    public long getIdLong() {
        return id;
    }

    @Override
    public boolean equals(Object object) {
        if (object == this) {
            return true;
        }
        if (!(object instanceof Role)) {
            return false;
        }
        Role role = (Role) object;
        return this.getIdLong() == role.getIdLong();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(id);
    }

    @Override
    public String toString() {
        return "R:" + getName() + '(' + id + ')';
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: src/main/java/me/syari/ss/discord/internal/entities/Role.kt
package me.syari.ss.discord.internal.entities

import me.syari.ss.discord.api.WithId

class Role(override val idLong: Long, val name: String): WithId {
    companion object {
        private val roleList = mutableListOf<Role>()

        fun add(role: Role) {
            roleList.add(role)
        }

        fun get(id: Long): Role? {
            return roleList.firstOrNull { it.idLong == id }
        }
    }

    init {
        add(this)
    }

    val asMention = "<@&$idLong>"

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other !is Role) return false
        return idLong == other.idLong
    }

    override fun hashCode(): Int {
        return java.lang.Long.hashCode(idLong)
    }

    override fun toString(): String {
        return "Role:$name($idLong)"
    }
}